include/zephyr/drivers/crc.h

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2024 Brill Power Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CRC_H
+#define ZEPHYR_INCLUDE_DRIVERS_CRC_H
+
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief CRC driver APIs
+ * @defgroup crc_interface CRC driver APIs
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#define CRC_FLAG_REVERSE_INPUT	BIT(0)
+#define CRC_FLAG_REVERSE_OUTPUT BIT(1)
+
+#define CRC4_INIT_VAL	     0x0
+#define CRC4_TI_INIT_VAL     0x0
+#define CRC7_BE_INIT_VAL     0x0
+#define CRC8_INIT_VAL	     0x0
+#define CRC8_CCITT_INIT_VAL  0x00
+#define CRC16_INIT_VAL	     0x0
+#define CRC16_ANSI_INIT_VAL  0x0
+#define CRC16_CCITT_INIT_VAL 0x0000
+#define CRC16_ITU_T_INIT_VAL 0x0000
+#define CRC24_PGP_INIT_VAL   0x0
+#define CRC32_C_INIT_VAL     0x00000000
+#define CRC32_IEEE_INIT_VAL  0xFFFFFFFFU
+
+#define CRC8_CCITT_POLY      0x07
+#define CRC32_IEEE_POLY      0x04C11DB7U
+
+/**
+ * @brief CRC algorithm enumeration
+ */
+enum crc_type {
+	CRC4,
+	CRC4_TI,
+	CRC7_BE,
+	CRC8,
+	CRC8_CCITT,
+	CRC16,
+	CRC16_ANSI,
+	CRC16_CCITT,
+	CRC16_ITU_T,
+	CRC24_PGP,
+	CRC32_C,
+	CRC32_IEEE,
+};
+
+/**
+ * @brief CRC state enumeration
+ */
+enum crc_state {
+	CRC_STATE_IDLE,
+	CRC_STATE_IN_PROGRESS
+};
+
+typedef uint64_t crc_init_val_t;
+typedef uint64_t crc_polynomial_t;
+typedef uint64_t crc_result_t;
+
+/**
+ * @brief CRC context structure
+ */
+struct crc_ctx {
+	enum crc_type type;		/**< Type of CRC algorithm */
+	enum crc_state state;		/**< State of CRC calculation */
+	crc_init_val_t initial_value;	/**< Initial value for CRC calculation */
+	uint32_t flags;			/**< Flags for CRC calculation */
+	crc_polynomial_t polynomial;	/**< Polynomial used in CRC calculation */
+	crc_result_t result;		/**< Store the CRC result */
+};
+
+typedef int (*crc_api_begin)(const struct device *dev, struct crc_ctx *ctx);
+typedef int (*crc_api_finish)(const struct device *dev, struct crc_ctx *ctx, const void *buffer,
+			      size_t bufsize);
```

Comment:
Discontinuous buffers can be processed by calling the finish function multiple times to process each buffer and retrieving the final result from ctx once everything has been processed. I recognize that in this scenario it wouldn't make much sense to call finish multiple times so maybe this function can be renamed to update and the finish function can be omitted all together. The final result would be able to be retrieved directly from the ctx once all updates are completed. 

Is there a particular use case you had in mind for hardware peripheral that uses discontinuous buffers?

Reply:
It is not possible to call `finish` multiple times in the current implementation:
https://github.com/zephyrproject-rtos/zephyr/blob/fac51a0f02e1cfb2dc7c122a98e813e982ef2e37/drivers/crc/crc_stm32.c#L103-L109

If several consecutive calls of `finish` is what you had in mind, then maybe it would be a better fit to have a single `crc_update` function.

Reply:
I went ahead and reintroduced a finish function to allow for multiple consecutive calls to update, indicate the calculation is complete, and release the semaphore.